### PR TITLE
Added WithProperty and SetProperty on Logger

### LINKED ISSUE #528: Feature request: define/set variable inside a logger
I would like to defined (or change) a variable for a particular logger.

For example:

``` xml
<?xml version="1.0" encoding="utf-8"?>

<nlog>
  <targets>
    <target name="db" xsi:type="Database" keepConnection="true" useTransactions="false"
            connectionString="server='...';user id='...';password='...';database='...'"
            commandText="">
      <parameter name="@message" layout="${message}" />
      <parameter name="@category" layout="${category}" />
    </target>
  </targets>
  <rules>
    <logger name="*.sqlExecutor" minlevel="Debug" writeTo="db">
      <!--new, variable for this logger-->
      <variable name="category" value="queries" />
    </logger>
  </rules>
</nlog>
```

### PR DESCRIPTION
**Pull request by snakefoot**
Logger - WithProperty for adding Logger specific properties that are included for all LogEvents

Different attempt of implementing #2496

Still trying to resolve #528

---
This change is [Reviewable](https://reviewable.io/reviews/nlog/nlog/3298)

### CODE REVIEW

**Comment by 304NotModified** [comment]
Thanks for the PR,

but i'm afraid it won't change the NLog API viewpoint, mentioned in https://github.com/NLog/NLog/pull/2496#issuecomment-410059676?

**Comment by snakefoot** [comment]
The Logger currently has name that matches the context it represents. For every context you create a new logger.


This PR allows you to inject more detail into the context that the Logger represents (besides its name).


I don't like global context properties that affects all users of the same logger. Kind of ruins the idea of context.


Sent from my Sony Xperia


---- Julian Verdurmen wrote ----


Thanks for the PR,

but i'm afraid it won't change the NLog API viewpoint, mentioned in #2496 (comment)<https://github.com/NLog/NLog/pull/2496#issuecomment-410059676>?

**Comment by 304NotModified** [comment]
If we also add SetProperty, next to WithProperty, do you think that is OK? (from viewpoint API design)

**Comment by snakefoot** [comment]
@304NotModified Added a `SetProperty` method for updating the current object-instance.

**Comment by 304NotModified** [comment]
Cool!

**Comment by 304NotModified** [comment]
Need to check it in depth, but I think I will merge it soon :)

File: src/NLog/Logger.cs [hunk 1]
```diff
@@ -94,17 +97,70 @@ public bool IsEnabled(LogLevel level)
...
+                : new Dictionary<string, object>();
+            contextProperties[propertyKey] = propertyValue;
+            newLogger._contextProperties = contextProperties;
+            newLogger._contextLogger = _contextLogger;
+            newLogger._isTraceEnabled = true;
+            newLogger._isDebugEnabled = true;
+            newLogger._isInfoEnabled = true;
+            newLogger._isWarnEnabled = true;
+            newLogger._isErrorEnabled = true;
+            newLogger._isFatalEnabled = true;
```
Comment (304NotModified) [inline src/NLog/Logger.cs]: Why is everything enabled for the new logger?

File: src/NLog/Logger.cs [hunk 2]
```diff
@@ -94,17 +97,70 @@ public bool IsEnabled(LogLevel level)
...
+        /// <param name="propertyValue">Property Value</param>
+        /// <returns>New Logger object that automatically appends specified property</returns>
+        public Logger WithProperty(string propertyKey, object propertyValue)
+        {
+            if (string.IsNullOrEmpty(propertyKey))
+                throw new ArgumentException(nameof(propertyKey));
+
+            Logger newLogger = Factory.CreateNewLogger(GetType()) ?? new Logger();
+            newLogger.Initialize(Name, _configuration, Factory);
+            var contextProperties = _contextProperties != null
```
Comment (304NotModified) [inline src/NLog/Logger.cs]: please create a helper for creating contextProperties  - this is prevent to get the code complexity above the boundary

File: src/NLog/Logger.cs [hunk 3]
```diff
@@ -94,17 +97,70 @@ public bool IsEnabled(LogLevel level)
...
+        /// </remarks>
+        /// <param name="propertyKey">Property Name</param>
+        /// <param name="propertyValue">Property Value</param>
+        public void SetProperty(string propertyKey, object propertyValue)
+        {
+            if (string.IsNullOrEmpty(propertyKey))
+                throw new ArgumentException(nameof(propertyKey));
+
+            var currentProperties = _contextProperties;
+            var contextProperties = new Dictionary<string, object>((currentProperties?.Count ?? 0) + 1);
```
Comment (304NotModified) [inline src/NLog/Logger.cs]: Could we reuse the helper for contextProperties  here? (see earlier comment)

File: src/NLog/Logger.cs [hunk 4]
```diff
@@ -557,58 +612,94 @@ internal void Initialize(string name, LoggerConfiguration loggerConfiguration, L
...
+            _isErrorEnabled = IsEnabled(LogLevel.Error);
+            _isFatalEnabled = IsEnabled(LogLevel.Fatal);
 
             OnLoggerReconfigured(EventArgs.Empty);
         }
 
         private TargetWithFilterChain GetTargetsForLevel(LogLevel level)
         {
-            return _configuration.GetTargetsForLevel(level);
+            if (ReferenceEquals(_contextLogger, this))
```
Comment (304NotModified) [inline src/NLog/Logger.cs]: this is a performance tweak isn't? Please add a code comment  (before it gets refactored ;))

File: src/NLog/LoggerImpl.cs [hunk 5]
```diff
@@ -51,15 +51,10 @@ internal static class LoggerImpl
         private const int StackTraceSkipMethods = 0;
 
         [System.Diagnostics.CodeAnalysis.SuppressMessage("Microsoft.Naming", "CA2204:Literals should be spelled correctly", Justification = "Using 'NLog' in message.")]
-        internal static void Write([NotNull] Type loggerType, TargetWithFilterChain targets, LogEventInfo logEvent, LogFactory factory)
+        internal static void Write([NotNull] Type loggerType, [NotNull] TargetWithFilterChain targetsForLevel, LogEventInfo logEvent, LogFactory factory)
```
Comment (304NotModified) [inline src/NLog/LoggerImpl.cs]: Resharper annotations, always +1

File: src/NLog/Logger.cs (hunk 1 above)
Comment (snakefoot) [reply src/NLog/Logger.cs]: Noise from my first attempt. Will remove it

File: src/NLog/Logger.cs (hunk 2 above)
Comment (snakefoot) [reply src/NLog/Logger.cs]: Will fix it

File: src/NLog/Logger.cs (hunk 3 above)
Comment (snakefoot) [reply src/NLog/Logger.cs]: Will fix it

File: src/NLog/Logger.cs (hunk 4 above)
Comment (snakefoot) [reply src/NLog/Logger.cs]: Nope avoid recursive loop.

File: src/NLog/Logger.cs [hunk 6]
```diff
@@ -94,17 +97,62 @@ public bool IsEnabled(LogLevel level)
...
+
         /// <summary>
         /// Writes the specified diagnostic message.
         /// </summary>
         /// <param name="logEvent">Log event.</param>
         public void Log(LogEventInfo logEvent)
         {
-            if (IsEnabled(logEvent.Level))
+            var targetsForLevel = IsEnabled(logEvent.Level) ? GetTargetsForLevel(logEvent.Level) : null;
+            if (targetsForLevel != null)
```
Comment (304NotModified) [inline src/NLog/Logger.cs]: is this a performance tweak?

File: src/NLog/Logger.cs [hunk 7]
```diff
@@ -392,25 +441,23 @@ public void Log<TArgument>(LogLevel level, [Localizable(false)] string message,
             }
         }
 
-        private void WriteToTargets(LogLevel level, Exception ex, [Localizable(false)] string message, object[] args)
```
Comment (304NotModified) [inline src/NLog/Logger.cs]: less overloads = 👍

**Comment by snakefoot** [comment]
Think it is safe for NLog 4.6.3<tel:4.6.3>


Not sure what you are referring to regarding "performance tweak"


Sent from my Sony Xperia


---- Julian Verdurmen wrote ----


@304NotModified approved this pull request.

Thanks!

Is this safe for 4.6.3?

________________________________

In src/NLog/Logger.cs<https://github.com/NLog/NLog/pull/3298#discussion_r275955654>:

>          /// <summary>

         /// Writes the specified diagnostic message.

         /// </summary>

         /// <param name="logEvent">Log event.</param>

         public void Log(LogEventInfo logEvent)

         {

-            if (IsEnabled(logEvent.Level))

+            var targetsForLevel = IsEnabled(logEvent.Level) ? GetTargetsForLevel(logEvent.Level) : null;

+            if (targetsForLevel != null)



is this a performance tweak?

________________________________

In src/NLog/Logger.cs<https://github.com/NLog/NLog/pull/3298#discussion_r275955760>:

> @@ -392,25 +441,23 @@ public void Log<TArgument>(LogLevel level, [Localizable(false)] string message,

             }

         }



-        private void WriteToTargets(LogLevel level, Exception ex, [Localizable(false)] string message, object[] args)



less overloads = 👍

**Comment by 304NotModified** [comment]
added here: https://github.com/NLog/NLog/wiki/EventProperties-Layout-Renderer

**Comment by snakefoot** [comment]
Updated the docs a little so it the difference between `WithProperty` and `SetProperty` is clearer. So one will not be caught by the trap-door created by `SetProperty`.

**Comment by 304NotModified** [comment]
It's not s trap but feature ;)

**Comment by 304NotModified** [comment]
Updated it with a 2nd log. The set is really useful is some cases (e.g when you get more context along the way)

**Comment by snakefoot** [comment]
> The set is really useful is some cases (e.g when you get more context along the way)

Not when having the global NLog Logger-cache where multiple locations are referencing the same Logger-object, and then one location suddenly decides to call SetProperty.

**Comment by 304NotModified** [comment]
It's not suddenly? It's by (logger) name.

**Comment by 304NotModified** [comment]
Buy anyway, I'm curious when we will implement WithProperties and ClearProperty :p

(Maybe nice with default interfaces - c# 8)

**Comment by snakefoot** [comment]
> It's not suddenly? It's by (logger) name.

Not sure I understand. There is no guarantee that two isolated locations calls `GetLogger()` with the same name. When one of the isolated locations calls `SetProperty` then it will affect the other.

**Pull request merged**